File: leetcode/p00133/p00133.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Node:
    def __init__(self, val=0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def cloneGraph(self, node: 'Node') -> 'Node':
        visited = {}
        head = Node()
        self.dfs(visited, node, head)

        return head
    
    def dfs(self, visited: dict, src: 'Node', dst: 'Node'):
        dst.val = src.val
        visited[src] = dst

        for node in src.neighbors:
            if node in visited:
                dst.neighbors.append(visited[node])
                continue
            
            new_node = Node()
            dst.neighbors.append(new_node)
            self.dfs(visited, node, new_node)

File: leetcode/p00133/p00133.go
package p00133

type Node struct {
	Val       int
	Neighbors []*Node
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func cloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	visited := map[*Node]*Node{}
	head := &Node{}
	dfs(visited, node, head)

	return head
}

func dfs(visited map[*Node]*Node, src, dst *Node) {
	dst.Val = src.Val
	visited[src] = dst

	for _, node := range src.Neighbors {
		if reference, exists := visited[node]; exists {
			dst.Neighbors = append(dst.Neighbors, reference)
			continue
		}

		new_node := &Node{}
		dst.Neighbors = append(dst.Neighbors, new_node)
		dfs(visited, node, new_node)
	}
}
